v2/pkg/engine/datasource/pubsub_datasource/pubsub_datasource.go

```diff
@@ -321,38 +326,113 @@ func (s *RequestDataSource) Load(ctx context.Context, input []byte, w io.Writer)
+				if !p.visitor.Definition.InputValueDefinitionNameBytes(inputValueDefinitionRef).Equals(inputValueNameBytes) {
+					continue
+				}
+				inputObjectDefinitionRef, lastInputValueTypeRef = p.inputObjectDefinitionRefByInputValueDefinitionRef(inputValueDefinitionRef)
+				break
+			}
+		}
+		if inputObjectDefinitionRef != -1 {
+			return "", fmt.Errorf(`the event subject "%s" defines the final nested input field "%s" in argument template #%d, but it is not a leaf type`, subject, argumentPath[len(argumentPath)-1], templateNumber+1)
+		}
+		if lastInputValueTypeRef == -1 {
+			return "", fmt.Errorf(`the event subject "%s" defines the final nested input field "%s" in argument template #%d, but it does not exist`, subject, argumentPath[len(argumentPath)-1], templateNumber+1)
+		}
+		argumentName := argumentPath[0]
+		argumentRef, ok := p.visitor.Operation.FieldArgument(fieldRef, []byte(argumentName))
+		if !ok {
+			return "", fmt.Errorf(`argument "%s" is not defined`, argumentName)
+		}
+		p.visitor.Definition.ObjectTypeDefinitionFieldWithName(p.visitor.Walker.EnclosingTypeDefinition.Ref, p.visitor.Operation.FieldNameBytes(fieldRef))
+		argumentValue := p.visitor.Operation.ArgumentValue(argumentRef)
+		if argumentValue.Kind != ast.ValueKindVariable {
+			return "", fmt.Errorf(`expected argument "%s" kind to be "ValueKindVariable" but received "%s"`, argumentName, argumentValue.Kind)
+		}
+		variableName := p.visitor.Operation.VariableValueNameBytes(argumentValue.Ref)
+		_, ok = p.visitor.Operation.VariableDefinitionByNameAndOperation(p.visitor.Walker.Ancestors[0].Ref, variableName)
+		if !ok {
+			return "", fmt.Errorf(`expected definition to exist for variable "%s"`, variableName)
+		}
+		// the variable path should be the variable name, e.g., "a", and then the 2nd element from the path onwards
+		variablePath := append([]string{string(variableName)}, argumentPath[1:]...)
+
+		// the definition is passed as operation because getJSONRootType resolves the type from the first argument,
+		// but lastInputValueTypeRef comes from the definition
+		renderer, err := resolve.NewPlainVariableRendererWithValidationFromTypeRef(p.visitor.Definition, p.visitor.Definition, lastInputValueTypeRef, variablePath...)
+		if err != nil {
+			return "", err
+		}
+		contextVariable := &resolve.ContextVariable{
+			Path:     variablePath,
+			Renderer: renderer,
+		}
+		// We need to replace the template literal with the variable placeholder (and reuse if it already exists)
+		variablePlaceHolder, _ := p.variables.AddVariable(contextVariable) // $$0$$
+		substitutedSubject = strings.ReplaceAll(substitutedSubject, matches[0], variablePlaceHolder)
 	}
-	contextVariable := &resolve.ContextVariable{
-		Path:     []string{string(variableName)},
-		Renderer: renderer,
+	// substitute the variable templates for dummy values to check that the string is a valid NATS subject
+	if natsserver.IsValidSubject(variableTemplateRegex.ReplaceAllLiteralString(substitutedSubject, "a")) {
```

Comment:
Can we abstract it in a way that we don't get provider specific in the pubsub implementation? For Kafka, topics should be immutable.